### PR TITLE
Unify hash functions for pairs

### PR DESCRIPTION
**Pull request by 4tXJ7f**
CVC4 was implementing multiple slightly different hash functions for
pairs. With pull request #1157, we have a decent generic hash function
for pairs. This commit replaces the existing hash function
implementations with a typedef of the generic hash function.

### CODE REVIEW

**Comment by 4tXJ7f** [comment]
Note: this PR includes PR #1157, so it should be merged after that.
Note 2: Performance results on 10'000 random SMT-LIB benchmarks:
```
         DIRECTORY |  STAT |  EXIT |    CPU[s] |  SPACE[MB] |
        log_master |    ok |   871 |  146520.2 |   613927.6 |
log_replace_hashes |    ok |   871 |  146538.5 |   612653.1 |
```

File: src/theory/ite_utilities.h [hunk 1]
```diff
@@ -240,14 +241,8 @@ class ITESimplifier {
...
-  struct NodePairHashFunction {
-    size_t operator () (const NodePair& pair) const {
-      size_t hash = 0;
-      hash = 0x9e3779b9 + NodeHashFunction().operator()(pair.first);
-      hash ^= 0x9e3779b9 + NodeHashFunction().operator()(pair.second) + (hash << 6) + (hash >> 2);
-      return hash;
-    }
-  };/* struct ITESimplifier::NodePairHashFunction */
+  typedef PairHashFunction<Node, Node, NodeHashFunction, NodeHashFunction>
+      NodePairHashFunction;
```
Comment (timothy-king) [inline src/theory/ite_utilities.h]: nitpick: Switch to "using NodePairHashFunction = PairHashFunction<Node, Node, NodeHashFunction, NodeHashFunction>;"?

Comment (4tXJ7f) [reply src/theory/ite_utilities.h]: Nice! I didn't know about that syntax. That is actually much more readable than `typedef` :)

File: src/expr/node.h [hunk 2]
```diff
@@ -37,13 +37,14 @@
 #include "base/cvc4_assert.h"
 #include "base/exception.h"
 #include "base/output.h"
-#include "expr/type.h"
-#include "expr/kind.h"
-#include "expr/metakind.h"
 #include "expr/expr.h"
 #include "expr/expr_iomanip.h"
+#include "expr/kind.h"
```
Comment (timothy-king) [inline src/expr/node.h]: I am going to be annoying and ask this be in separate commit. node.h and these are mutually recursive and hence a bit flaky to re-ordering. This should be fine and should be done. Just I would like to be able to roll this back if we find it stops compiling on some obscure path.

Comment (4tXJ7f) [reply src/expr/node.h]: Good point, reverted the reordering.

**Pull request merged**